middle_end/flambda2/types/grammar/type_grammar.ml

```diff
@@ -1171,6 +1171,286 @@ let apply_coercion t coercion =
+        ~free_names_head:free_names_head_of_kind_value
+    in
+    if ty == ty' then t else Value ty'
+  | Naked_immediate ty ->
+    let ty' =
+      TD.remove_unused_closure_vars ty ~used_closure_vars
+        ~remove_unused_closure_vars_head:
+          remove_unused_closure_vars_head_of_kind_naked_immediate
+        ~apply_renaming_head:apply_renaming_head_of_kind_naked_immediate
+        ~free_names_head:free_names_head_of_kind_naked_immediate
+    in
+    if ty == ty' then t else Naked_immediate ty'
+  | Naked_float ty ->
+    let ty' =
+      TD.remove_unused_closure_vars ty ~used_closure_vars
+        ~remove_unused_closure_vars_head:
+          remove_unused_closure_vars_head_of_kind_naked_float
+        ~apply_renaming_head:apply_renaming_head_of_kind_naked_float
+        ~free_names_head:free_names_head_of_kind_naked_float
+    in
+    if ty == ty' then t else Naked_float ty'
+  | Naked_int32 ty ->
+    let ty' =
+      TD.remove_unused_closure_vars ty ~used_closure_vars
+        ~remove_unused_closure_vars_head:
+          remove_unused_closure_vars_head_of_kind_naked_int32
+        ~apply_renaming_head:apply_renaming_head_of_kind_naked_int32
+        ~free_names_head:free_names_head_of_kind_naked_int32
+    in
+    if ty == ty' then t else Naked_int32 ty'
+  | Naked_int64 ty ->
+    let ty' =
+      TD.remove_unused_closure_vars ty ~used_closure_vars
+        ~remove_unused_closure_vars_head:
+          remove_unused_closure_vars_head_of_kind_naked_int64
+        ~apply_renaming_head:apply_renaming_head_of_kind_naked_int64
+        ~free_names_head:free_names_head_of_kind_naked_int64
+    in
+    if ty == ty' then t else Naked_int64 ty'
+  | Naked_nativeint ty ->
+    let ty' =
+      TD.remove_unused_closure_vars ty ~used_closure_vars
+        ~remove_unused_closure_vars_head:
+          remove_unused_closure_vars_head_of_kind_naked_nativeint
+        ~apply_renaming_head:apply_renaming_head_of_kind_naked_nativeint
+        ~free_names_head:free_names_head_of_kind_naked_nativeint
+    in
+    if ty == ty' then t else Naked_nativeint ty'
+  | Rec_info ty ->
+    let ty' =
+      TD.remove_unused_closure_vars ty ~used_closure_vars
+        ~remove_unused_closure_vars_head:
+          remove_unused_closure_vars_head_of_kind_rec_info
+        ~apply_renaming_head:apply_renaming_head_of_kind_rec_info
+        ~free_names_head:free_names_head_of_kind_rec_info
+    in
+    if ty == ty' then t else Rec_info ty'
+
+and remove_unused_closure_vars_head_of_kind_value head ~used_closure_vars =
+  match head with
+  | Variant { blocks; immediates; is_unique } ->
+    let immediates' =
+      let>+ immediates = immediates in
+      remove_unused_closure_vars immediates ~used_closure_vars
+    in
+    let blocks' =
+      let>+ blocks = blocks in
+      remove_unused_closure_vars_row_like_for_blocks blocks ~used_closure_vars
+    in
+    if immediates == immediates' && blocks == blocks'
```

Comment:
`Or_unknown.(let>+)` doesn't preserve sharing currently, so this equality will only hold for variants with `Unknown` blocks and immediates. Maybe we should be using `map_sharing` instead of `map` for `let>+` ?